package/piksi_system_daemon/piksi_system_daemon/src/cellmodem_probe.c

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (C) 2017 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <termios.h>
+#include <unistd.h>
+#include <stdarg.h>
+
+#include <libpiksi/sbp_zmq_pubsub.h>
+#include <libpiksi/logging.h>
+#include <libsbp/logging.h>
+
+#include "cellmodem.h"
+#include "cellmodem_probe.h"
+
+enum {
+  RESPONSE_OK,
+  RESPONSE_ERROR,
+};
+
+static const char *const response_strs[] = {
+  [RESPONSE_OK] = "OK\r\n",
+  [RESPONSE_ERROR] = "ERROR\r\n",
+  NULL
+};
+
+static int modem_read(int fd, char *buf, size_t len, const char *const *responses)
+{
+  int n = 0;
+  fd_set fds;
+  struct timeval timeout = {.tv_sec = 0, .tv_usec = 100000};
+  do {
+    FD_ZERO(&fds);
+    FD_SET(fd, &fds);
+    if (select(fd+1, &fds, NULL, NULL, &timeout) < 0)
+      return -2;
+    if (!FD_ISSET(fd, &fds)) {
+      return -1;
+    }
+    int r = read(fd, buf + n, len - n - 1);
+    if (r < 0) {
+        return -3;
+    }
+    n += r;
+    buf[n] = 0;
+    for(int i = 0; responses[i]; i++) {
+      int rlen = strlen(responses[i]);
+      if ((n >= rlen) && (strcmp(buf + n - rlen, responses[i]) == 0)) {
+        buf[n - rlen] = 0;
+        return i;
+      }
+    }
+  } while (n < len - 1);
+}
+
+#define DELIM "\r\n"
+static int cellmodem_command(int fd, const char *cmd, char *response, size_t len)
+{
+  /* Send command */
+  write(fd, cmd, strlen(cmd));
+  write(fd, "\r", 1);
+
+  /* Read full response */
+  char buf[256] = {0};
+  if (modem_read(fd, buf, sizeof(buf), response_strs) != RESPONSE_OK)
+    return -1;
+
+  /* Check command echo */
+  if (strncmp(buf, cmd, strlen(cmd)) != 0)
+    return -2;
+
+  /* Pull out response */
+  if (response) {
+    /* Strip \r\n junk */
+    int i;
+    for (i = strlen(buf) - 1; (buf[i] == '\r') || (buf[i] == '\n'); i--)
+      buf[i] = 0;
+    for (i = strlen(cmd); (buf[i] == '\r') || (buf[i] == '\n'); i++)
+      ;
+
+    strncpy(response, buf + i, len);
+  }
+
+  return 0;
+}
+
+static void log_info(sbp_zmq_pubsub_ctx_t *pubsub_ctx, const char *fmt, ...)
```

Comment:
You can use the `sbp_log` helper for this.

Reply:
Thanks, I'll do that!